### PR TITLE
[#289] Handle nomatch for string:prefix/2

### LINKED ISSUE #289: Elvis Diagnostics cause server crash
**Describe the bug**

```
{badarg,[{erlang,binary_to_list,[nomatch],[]},{els_utils,project_relative,1,[{file,"/.../erlang_ls/src/els_utils.erl"},{line,77}]},{els_elvis_diagnostics,diagnostics,1,[{file,"/.../erlang_ls/src/els_elvis_diagnostics.erl"},{line,26}]},{els_text_synchronization,did_save,2,[{file,"/.../erlang_ls/src/els_text_synchronization.erl"},{line,22}]}]}
```

### PR DESCRIPTION
**Pull request by jfacorro**
### Description

Handle `nomatch` result for `string:prefix/2`. Depending on what the client provides as a `root_uri`, the value might or might not include a trailing `/`, we should handle both cases.

Fixes #289.

### CODE REVIEW

File: src/els_utils.erl [hunk 1]
```diff
@@ -70,17 +70,23 @@ start_epmd() ->
...
 -spec project_relative(uri()) -> file:filename().
 project_relative(Uri) ->
-  Path = els_uri:path(Uri),
-  RootDir = els_config:get(root_uri),
-  RootPath = els_uri:path(RootDir),
-  RelativeBin = string:prefix(string:prefix(Path, RootPath), <<"/">>),
+  Path        = els_uri:path(Uri),
+  RootUri     = els_config:get(root_uri),
+  RootPath    = els_uri:path(RootUri),
+  RelativeBin = prefix(prefix(Path, RootPath), <<"/">>),
```
Comment (robertoaloi) [inline src/els_utils.erl]: Uhm, this will not work on Windows, will it?

Comment (alanz) [reply src/els_utils.erl]: My previous version had the benefit of using file-system aware split and join.

But a colleague told me Erlang works with `/` as a path separator internally, but then switches it when it hits the file system.  Which won't work if the URL comes from non-Erlang.

Comment (alanz) [reply src/els_utils.erl]: Maybe we should revert to my original version? We have an internal utility function now anyway, so no real saving.

And at least it intentionally operates on file paths, rather than strings.

Comment (jfacorro) [reply src/els_utils.erl]: But aren't URIs file system independent and always expressed using `/`?

Comment (robertoaloi) [reply src/els_utils.erl]: What about something like:

```
%% @doc Returns a project-relative file path for a given URI
-spec project_relative(uri()) -> file:filename().
project_relative(Uri) ->
  RootUri     = els_config:get(root_uri),
  RootUriSize = byte_size(RootUri),
  <<RootUri:RootUriSize/binary, RelativePath/binary>> = Uri,
  binary_to_list(filename:basename(RelativePath)).
```

Comment (robertoaloi) [reply src/els_utils.erl]: Except that does not seem to work, let me try again... :)

**Comment by jfacorro** [comment]
I tried adding a build in a windows machine but AppVeyor only includes OTP 20 in their Windows images 😞.

File: src/els_utils.erl (hunk 1 above)
Comment (jfacorro) [reply src/els_utils.erl]: > Maybe we should revert to my original version? We have an internal utility function now anyway, so no real saving.
> 
> And at least it intentionally operates on file paths, rather than strings.

That is a good point 😜. I'm open to using the original version.

Comment (robertoaloi) [reply src/els_utils.erl]: Btw, the fact this exists is not comforting...

https://github.com/erlang/rebar3/blob/master/src/rebar_file_utils.erl#L429

Comment (robertoaloi) [reply src/els_utils.erl]:
```
  RootUri     = els_config:get(root_uri),
  RootUriSize = byte_size(RootUri),
  <<RootUri:RootUriSize/binary, RelativePath/binary>> = Uri,
  binary_to_list(string:trim(RelativePath, leading, [$/, $\\])).
```

**Comment by robertoaloi** [comment]
@jfacorro That would have actually been nice.

**Comment by robertoaloi** [comment]
@alanz I proposed an alternative solution. Whatever you guys prefer, I'm happy with.
